pulumitest/run.go

```diff
@@ -0,0 +1,95 @@
+package pulumitest
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/pulumi/providertest/pulumitest/optrun"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/apitype"
+)
+
+// Run will run the `execute` function in an isolated temp directory and with additional test options, then import the resulting stack state into the original test.
+// WithCache can be used to skip executing the run and return the cached stack state if available, or to cache the stack state after executing the run.
+// Options will be inherited from the original test, but can be added to with `optrun.WithOpts` or reset with `opttest.Defaults()`.
+func (pulumiTest *PulumiTest) Run(execute func(test *PulumiTest), opts ...optrun.Option) *PulumiTest {
+	pulumiTest.T().Helper()
+
+	options := optrun.DefaultOptions()
+	for _, o := range opts {
+		o.Apply(options)
+	}
+
+	var stackExport *apitype.UntypedDeployment
+	var err error
+	if options.EnableCache {
+		stackExport, err = tryReadStackExport(options.CachePath)
+		if err != nil {
+			pulumiTest.T().Fatalf("failed to read stack export: %v", err)
+		}
+		if stackExport != nil {
+			pulumiTest.T().Logf("run cache found at %s", options.CachePath)
+		} else {
+			pulumiTest.T().Logf("no run cache found at %s", options.CachePath)
+		}
+	}
+
+	if stackExport == nil {
+		isolatedTest := pulumiTest.CopyToTempDir(options.OptTest...)
+		execute(isolatedTest)
+		isolatedTest.T().Logf("writing stack state to %s", options.CachePath)
```

Comment:
That's only true if `options.EnableCache` is true, isn't it?

Reply:
Yes!